### PR TITLE
ESG

### PR DESCRIPTION
**Pull request by Kazeya27**

### CODE REVIEW

**Comment by Kazeya27** [comment]
原论文用的是2016年4月到6月的NYC-bike/taxi数据集，我在云盘里没找到。

- 原论文

Dataset | Horizon-3 | Horizon-6 | Horizon-12
-- | -- | -- | --
NYC-Bike | RMSE=2.5529--MAE=1.5483 | RMSE=2.6484--MAE=1.6026|RMSE=2.8778--MAE=1.7173
NYC-Taxi | RMSE=8.5745--MAR=4.8750 | RMSE=9.0125--MAE=5.0500|RMSE=9.7857--MAE=5.4019

- libcity
Dataset | Horizon-3 | Horizon-6 | Horizon-12
-- | -- | -- | --
NYCBike20160809 | RMSE=4.823869--MAE=2.041650 | RMSE=4.975906--MAE=2.106852|RMSE=5.341728--MAE=2.222684
NYCTaxi20160102 | RMSE=12.480914--MAE=4.200171 | RMSE=13.343316--MAE=4.392735|RMSE=14.703619--MAE=4.710590

**Comment by Kazeya27** [comment]
参数与论文中用于NYC数据集的一致
{
  "max_epoch": 500,

  "learner": "adam",
  "learning_rate": 0.001,
  "lr_epsilon": 1e-8,
  "weight_decay": 0.0001,

  "clip_grad_norm": true,
  "max_grad_norm": 5,
  "use_early_stop": true,
  "patience": 30,

  "n_layers": 3,
  "dropout": 0.3,
  "residual_channels": 32,
  "end_channels": 128,
  "n_blocks": 1,
  "st_embedding_dim": 40,
  "kernel_set": [2, 6],
  "dilation_exp": 1,
  "conv_channels": 32,
  "gcn_depth": 2,
  "dy_embedding_dim": 20,
  "skip_channels": 64,
  "layer_norm_affline": false,
  "propalpha": 0.05,
  "dy_interval": [1, 1, 1]
}

{
  "batch_size": 16,
  "cache_dataset": false,
  "num_workers": 0,
  "pad_with_last_sample": true,
  "train_rate": 0.7,
  "eval_rate": 0.15,
  "scaler": "standard",
  "load_external": false,
  "normal_external": false,
  "ext_scaler": "none",
  "input_window": 12,
  "output_window": 12,
  "add_time_in_day": false,
  "add_day_in_week": false,
  "use_row_column": false
}

File: libcity/model/traffic_flow_prediction/ESG.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+class ESG(AbstractTrafficStateModel):
+    def __init__(self, config, data_feature):
+        super().__init__(config, data_feature)
+        self.device = config.get('device', torch.device('cpu'))
+        # section 1: data_feature
+        self._scaler = self.data_feature.get('scaler')
+        self.num_nodes = self.data_feature.get('num_nodes', 1)
+        self.feature_dim = self.data_feature.get('feature_dim', 1)
+        self.output_dim = self.data_feature.get('output_dim', 1)
+        torch.Tensor
```
Comment (NickHan-cs) [inline libcity/model/traffic_flow_prediction/ESG.py:303]: 这里是不是多出来了

File: libcity/model/traffic_flow_prediction/ESG.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+        in_channels = self.skip_channels
+        final_channels = self.pred_len * self.output_dim
+
+        self.out = nn.Sequential(
+            nn.ReLU(),
+            nn.Conv2d(in_channels, self.end_channels, kernel_size=(1, 1), bias=True),
+            nn.ReLU(),
+            nn.Conv2d(self.end_channels, final_channels, kernel_size=(1, 1), bias=True)
+        )
+        self.fc_dim = (self.static_feat.shape[0] - 18) * 16
```
Comment (NickHan-cs) [inline libcity/model/traffic_flow_prediction/ESG.py:358]: 这里的18和16硬编码是什么含义？

File: libcity/data/dataset/dataset_subclass/esg_dataset.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+import os
+import numpy as np
+
+from libcity.data.dataset import TrafficStateGridDataset
+
+
+class ESGDataset(TrafficStateGridDataset):
```
Comment (NickHan-cs) [inline libcity/data/dataset/dataset_subclass/esg_dataset.py:7]: 需要用GridDataset吗？看文章是用Graph形式输入的？

File: libcity/data/dataset/dataset_subclass/__init__.py [hunk 4]
```diff
@@ -46,6 +47,7 @@
     "GeoSANDataset",
     "GSNetDataset",
     "LINEDataset",
-    "STGODEDataset"
-    "DMSTGCNDataset"
+    "STGODEDataset",
+    "DMSTGCNDataset",
+    "ESGDataset"
```
Comment (NickHan-cs) [inline libcity/data/dataset/dataset_subclass/__init__.py:52]: 建议以后这里最后一行也加上逗号，这样别人改代码不会发生和你的修改部分发生冲突，model中的__init__同理

File: libcity/data/dataset/dataset_subclass/esg_dataset.py (hunk 3 above)
Comment (Kazeya27) [reply libcity/data/dataset/dataset_subclass/esg_dataset.py:8]: 这个是因为NYC-Bike是.grid文件

**Review by NickHan-cs — APPROVED** [review]

**Review by NickHan-cs — APPROVED** [review]
